panther_docking/include/panther_docking/panther_charging_dock.hpp

```diff
@@ -0,0 +1,243 @@
+
+  /**
+   * @brief Offset the staging pose.
+   *
+   * This method offsets the staging dock pose by values described in a configuration.
+   *
+   * @param dock_pose The dock pose to offset.
+   *
+   * @return The offset staging pose.
+   */
+  geometry_msgs::msg::PoseStamped offsetStagingPoseToDockPose(
+    const geometry_msgs::msg::PoseStamped & dock_pose);
+
+  /**
+   * @brief Offset the detected dock pose.
+   *
+   * This method offsets the detected dock pose by values described in a configuration.
+   *
+   * @param detected_dock_pose The detected dock pose to offset.
+   *
+   * @return The offset detected dock pose.
+   */
+  geometry_msgs::msg::PoseStamped offsetDetectedDockPose(
+    const geometry_msgs::msg::PoseStamped & detected_dock_pose);
+
+  /**
+   * @brief Get the dock pose in a detection dock frame.
+   *
+   * This method retrieves the dock pose in a detection dock frame.
+   *
+   * @param frame The detection frame to get the dock pose in.
+   * @return The dock pose in the detection frame.
+   */
+  geometry_msgs::msg::PoseStamped getDockPose(const std::string & frame);
+
+  /**
+   * @brief Method to update the dock pose and publish it.
+   *
+   * This method makes all necessary transformations to update the dock pose and publishes it.
+   *
+   */
+  void updateDockPoseAndPublish();
+
+  /**
+   * @brief Update the staging pose and publish it.
+   *
+   * This method makes all necessary transformations to update the staging pose and publishes it.
+   *
+   * @param frame The frame to publish the staging pose in.
+   */
+  void updateStagingPoseAndPublish(const std::string & frame);
+
+  std::string name_;
+  std::string base_frame_name_;
+  std::string dock_frame_;
+  double panther_version_;
+
+  rclcpp::Logger logger_{rclcpp::get_logger("PantherChargingDock")};
+  rclcpp::Clock steady_clock_{RCL_STEADY_TIME};
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node_;
+  std::shared_ptr<tf2_ros::Buffer> tf2_buffer_;
+  rclcpp::Subscription<panther_msgs::msg::ChargingStatus>::SharedPtr charging_status_sub_;
+  rclcpp::Subscription<panther_msgs::msg::IOState>::SharedPtr io_state_sub_;
+  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr staging_pose_pub_;
+  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr dock_pose_pub_;
+  rclcpp::Client<std_srvs::srv::SetBool>::SharedPtr charger_enable_client_;
+
+  realtime_tools::RealtimeBox<std::shared_ptr<panther_msgs::msg::ChargingStatus>>
+    charging_status_box_{nullptr};
+  realtime_tools::RealtimeBox<std::shared_ptr<panther_msgs::msg::IOState>> io_state_box_{nullptr};
+
+  geometry_msgs::msg::PoseStamped dock_pose_;
+  geometry_msgs::msg::PoseStamped staging_pose_;
+
+  double external_detection_timeout_;
+  tf2::Quaternion external_detection_rotation_;
+  double external_detection_translation_x_;
+  double external_detection_translation_y_;
+  double external_detection_translation_z_;
```

Comment:
If not struct, `tf2::Vector3` is an option as well since eventually those values are used to create this type.

I highlight this aspect because having a lot of primitive values as class members makes this class a bit cumbersome. This is a chance to containerize primitives since they naturally are in the same domain.

Reply:
I would leave this without changes because it should not be used in a product version.